lib/wserver/wserver.ml

```diff
@@ -258,97 +258,73 @@ let check_stopping () =
     flush stderr;
     exit 0)
 
+let client_connection tmout callback addr t =
+  let oc = Unix.out_channel_of_descr t in
+  connection_closed := false;
+  wserver_sock := t;
+  wserver_oc := oc;
+  Fun.protect ~finally:close_connection @@ fun () ->
+  treat_connection tmout callback addr t
+
 let accept_connection tmout max_clients callback s =
   let () = wait_available max_clients s in
   let t, addr = Unix.accept s in
-  check_stopping ();
-  Unix.setsockopt t Unix.SO_KEEPALIVE true;
-  if Sys.unix then (
-    if !no_fork then (
-      connection_closed := false;
-      wserver_sock := t;
-      wserver_oc := Unix.out_channel_of_descr t;
-      treat_connection tmout callback addr t;
-      close_connection ())
-    else
-      match try Some (Unix.fork ()) with _ -> None with
-      | Some 0 -> (
-          try
-            if max_clients = None && Unix.fork () <> 0 then exit 0;
-            Unix.close s;
-            wserver_sock := t;
-            wserver_oc := Unix.out_channel_of_descr t;
-            treat_connection tmout callback addr t;
-            close_connection ();
-            exit 0
-          with
-          | Unix.Unix_error (Unix.ECONNRESET, "read", _) -> exit 0
-          | e -> raise e)
-      | Some id ->
-          Unix.close t;
-          if max_clients = None then
-            let _ = Unix.waitpid [] id in
-            ()
-          else pids := id :: !pids
-      | None ->
-          Unix.close t;
-          eprintf "Fork failed\n";
-          flush stderr)
-  else
-    let oc = open_out_bin !sock_in in
-    let cleanup () = try close_out oc with _ -> () in
-    (try copy_what_necessary t oc with
-    | Unix.Unix_error (_, _, _) -> ()
-    | exc ->
-        cleanup ();
-        raise exc);
-    cleanup ();
-    (let pid =
-       let env =
-         Array.append (Unix.environment ())
-           [| "WSERVER=" ^ string_of_sockaddr addr |]
-       in
-       let args = Sys.argv in
-       Unix.create_process_env Sys.argv.(0) args env Unix.stdin Unix.stdout
-         Unix.stderr
-     in
-     let _ = Unix.waitpid [] pid in
-     let ic = open_in_bin !sock_in in
-     close_in ic);
```

Comment:
I don't understand the purpose of these lines. We open an input channel, then close it immediately. I kept it but I believe that it's pointless.